### ISSUE TITLE
Regression: SIGINT not handled correctly in new watchexec version

### ISSUE THREAD

**Issue by kurtbuilds** [watchexec/watchexec#247]
On old version of watchexec-cli, when I used Ctrl+C to send SIGINT, it would shut down the watchexec process and clean up the child process.

On the latest version (I assume since 1.18), this behavior doesn't work. I didn't investigate deeply what happens to the signal, but in my case, neither watchexec nor the child process appear to respond to SIGINT when I press Ctrl+C.

Right now, I stop the process (Ctrl+Z), and then use `kill`, but it seems like that approach leaves a process around, using my desired port (I'm running a local dev server).

I'd like to use the updated version, however this regression prevents doing so.

**Comment by passcod** [comment]
OS? Logs with `-vvv`? Command you're running?

**Comment by kurtbuilds** [comment]
Console output:

```
$ watchexec -e rs just run
cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/server2`
Listening on 0.0.0.0:5000...
Error: Os { code: 48, kind: AddrInUse, message: "Address already in use" }
error: Recipe `run` failed on line 11 with exit code 1
^C^C^C^C^C^C^C^C^Z
[1]+  Stopped                 watchexec -e rs just run
```

OS Mac OS X 11.6.2, watchexec-cli 1.18.3.

`-vvv` output:


```
C2022-01-25T18:50:50.034173Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-25T18:50:50.034194Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
2022-01-25T18:50:50.034208Z TRACE watchexec::action: got event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
2022-01-25T18:50:50.034250Z TRACE filterer_check: watchexec::filter::globset: new
2022-01-25T18:50:50.034268Z TRACE filterer_check: watchexec::filter::globset: checking internal ignore filterer
2022-01-25T18:50:50.034274Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: new
2022-01-25T18:50:50.034280Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: verdict pass=true
2022-01-25T18:50:50.034288Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: close time.busy=7.64µs time.idle=6.73µs
2022-01-25T18:50:50.034299Z TRACE filterer_check: watchexec::filter::globset: close time.busy=32.2µs time.idle=17.1µs
2022-01-25T18:50:50.034307Z TRACE watchexec::action: filter rejected event
2022-01-25T18:50:50.034330Z TRACE watchexec::action: nothing in set, waiting forever for next event
2022-01-25T18:50:50.034336Z TRACE watchexec::action: waiting for event maxtime=18446744073709551615s
```

**Comment by passcod** [comment]
It's in 1.18.4.

**Comment by kurtbuilds** [comment]
Thank you for updating this! The base mechanic is fixed for me now. However, there's a lingering issue where watchexec is reporting an error, when it should not.

When I run the command, `watchexec -e rs just run`, and then Ctrl+C to SIGINT, it prints this to console before exiting:

```
[[io(waiting on process group): No child processes (os error 10)]]
```

I believe this line is coming from watchexec, since Ctrl+C from `just run` prints nothing upon exit. I think it's important that tools not print out errors in the "standard course of business".

The trace is below:

```
^C2022-01-26T17:40:26.758663Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T17:40:26.758685Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
2022-01-26T17:40:26.758709Z TRACE watchexec::action: got event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
2022-01-26T17:40:26.758745Z TRACE filterer_check: watchexec::filter::globset: new
2022-01-26T17:40:26.758759Z TRACE filterer_check: watchexec::filter::globset: checking internal ignore filterer
2022-01-26T17:40:26.758767Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: new
2022-01-26T17:40:26.758775Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: verdict pass=true
2022-01-26T17:40:26.758784Z TRACE filterer_check:filterer_check: watchexec::ignore::filter: close time.busy=10.2µs time.idle=7.08µs
2022-01-26T17:40:26.758795Z TRACE filterer_check: watchexec::filter::globset: non-path event (pass)
2022-01-26T17:40:26.758800Z TRACE filterer_check: watchexec::filter::globset: close time.busy=41.4µs time.idle=14.4µs
2022-01-26T17:40:26.758807Z TRACE watchexec::action: filter passed event
2022-01-26T17:40:26.758824Z TRACE watchexec::action: event is the first, resetting throttle window
2022-01-26T17:40:26.758830Z TRACE watchexec::action: still within throttle window, cycling elapsed=118ns
2022-01-26T17:40:26.758836Z TRACE watchexec::action: waiting for event maxtime=99.994225ms
2022-01-26T17:40:26.863100Z TRACE watchexec::action: timed out, cycling
2022-01-26T17:40:26.863123Z TRACE watchexec::action: out of throttle on recycle
2022-01-26T17:40:26.863127Z TRACE watchexec::action: out of throttle, starting action process
2022-01-26T17:40:26.863143Z DEBUG watchexec::action: action constructed action=Action { events: [Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }], outcome: OnceCell(Uninit) }
2022-01-26T17:40:26.863153Z DEBUG watchexec::action: running action handler
2022-01-26T17:40:26.863214Z DEBUG watchexec::action: handler finished outcome=Both(Stop, Exit)
2022-01-26T17:40:26.863224Z TRACE watchexec::command::supervisor: supervisor state ongoing=true
2022-01-26T17:40:26.863230Z DEBUG watchexec::action: outcome resolved outcome=Both(Stop, Exit)
2022-01-26T17:40:26.863241Z TRACE watchexec::action: applying outcome outcome=Both(Stop, Exit)
2022-01-26T17:40:26.863246Z TRACE watchexec::action: applying outcome outcome=Stop
2022-01-26T17:40:26.863251Z TRACE watchexec::command::supervisor: sending kill intervention
2022-01-26T17:40:26.863256Z DEBUG watchexec::command::supervisor: waiting on supervisor completion
2022-01-26T17:40:26.863265Z DEBUG watchexec::command::process: killing process group pgid=Some(13202)
2022-01-26T17:40:26.863355Z TRACE watchexec::command::process: waiting on process group
2022-01-26T17:40:26.863753Z ERROR watchexec::command::supervisor: while waiting on process err=io(waiting on process group): No child processes (os error 10)
2022-01-26T17:40:26.863797Z TRACE watchexec::command::supervisor: marking process as done
2022-01-26T17:40:26.863802Z TRACE watchexec::command::supervisor: closing supervisor task early
2022-01-26T17:40:26.863818Z DEBUG watchexec::command::supervisor: supervisor completed
2022-01-26T17:40:26.863831Z TRACE watchexec::action: applying outcome outcome=Exit
2022-01-26T17:40:26.863840Z TRACE watchexec::action: nothing in set, waiting forever for next event
2022-01-26T17:40:26.863844Z TRACE watchexec::action: waiting for event maxtime=18446744073709551615s
2022-01-26T17:40:26.863862Z ERROR watchexec::watchexec: runtime error err=io(waiting on process group): No child processes (os error 10)
[[io(waiting on process group): No child processes (os error 10)]]
2022-01-26T17:40:26.863898Z TRACE watchexec::watchexec: got graceful exit request via runtime error, upgrading to crit
2022-01-26T17:40:26.863916Z TRACE watchexec::watchexec: got graceful exit request via critical error, erasing the error
2022-01-26T17:40:26.863921Z DEBUG watchexec::watchexec: main task graceful exit
```

**Comment by kurtbuilds** [comment]
Actually, I'm getting inconsistent behavior. Sometimes it's working as expected. Other times, it's failing to process any events at all. 

Mac OS 11.6.2, watchexec-cli 1.18.4

Command: `watchexec -vvv -e rs just run`

I'm editing a rust project using IntelliJ Ultimate. When I edit a file, here is the trace:

```
2022-01-26T18:00:55.502089Z TRACE watchexec::fs: receiving possible event from watcher event=Ok(Event { kind: Modify(Metadata(Any)), paths: ["/Users/kurt/work/rust-service/server/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None })
2022-01-26T18:00:55.502202Z TRACE watchexec::fs: processed notify event into watchexec event event=Event { tags: [Source(Filesystem), FileEventKind(Modify(Metadata(Any))), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs", file_type: Some(File) }], metadata: {} }
2022-01-26T18:00:55.502222Z TRACE watchexec::fs: receiving possible event from watcher event=Ok(Event { kind: Modify(Data(Content)), paths: ["/Users/kurt/work/rust-service/server/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None })
2022-01-26T18:00:55.502274Z TRACE watchexec::fs: processed notify event into watchexec event event=Event { tags: [Source(Filesystem), FileEventKind(Modify(Data(Content))), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs", file_type: Some(File) }], metadata: {} }
2022-01-26T18:00:55.502285Z TRACE watchexec::fs: receiving possible event from watcher event=Ok(Event { kind: Create(File), paths: ["/Users/kurt/work/rust-service/server/src/main.rs~"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None })
2022-01-26T18:00:55.502331Z  WARN watchexec::fs: failed to canonicalise event path err=Os { code: 2, kind: NotFound, message: "No such file or directory" } path="/Users/kurt/work/rust-service/server/src/main.rs~"
2022-01-26T18:00:55.502344Z TRACE watchexec::fs: processed notify event into watchexec event event=Event { tags: [Source(Filesystem), FileEventKind(Create(File)), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs~", file_type: None }], metadata: {} }
2022-01-26T18:00:55.502351Z TRACE watchexec::fs: receiving possible event from watcher event=Ok(Event { kind: Remove(File), paths: ["/Users/kurt/work/rust-service/server/src/main.rs~"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None })
2022-01-26T18:00:55.502395Z  WARN watchexec::fs: failed to canonicalise event path err=Os { code: 2, kind: NotFound, message: "No such file or directory" } path="/Users/kurt/work/rust-service/server/src/main.rs~"
2022-01-26T18:00:55.502404Z TRACE watchexec::fs: processed notify event into watchexec event event=Event { tags: [Source(Filesystem), FileEventKind(Remove(File)), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs~", file_type: None }], metadata: {} }
2022-01-26T18:00:55.502411Z TRACE watchexec::fs: receiving possible event from watcher event=Ok(Event { kind: Modify(Data(Content)), paths: ["/Users/kurt/work/rust-service/server/src/main.rs~"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None })
2022-01-26T18:00:55.502453Z  WARN watchexec::fs: failed to canonicalise event path err=Os { code: 2, kind: NotFound, message: "No such file or directory" } path="/Users/kurt/work/rust-service/server/src/main.rs~"
2022-01-26T18:00:55.510792Z TRACE watchexec::fs: processed notify event into watchexec event event=Event { tags: [Source(Filesystem), FileEventKind(Modify(Data(Content))), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs~", file_type: None }], metadata: {} }
```

And when I then tried to Ctrl+C from the process, I got this trace:

```
^C2022-01-26T18:01:50.925404Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:50.925424Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:51.109427Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:51.109448Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:51.580767Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:51.580789Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:52.369960Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:52.369985Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:52.558729Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:52.558774Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:52.758848Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:52.758869Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:01:52.874087Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:01:52.874107Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
^C2022-01-26T18:03:01.163112Z DEBUG watchexec::signal::source: received unix signal sig=Interrupt
2022-01-26T18:03:01.163132Z TRACE watchexec::signal::source: processed signal into event event=Event { tags: [Source(Keyboard), Signal(Interrupt)], metadata: {} }
```

In neither case did it re-run the command, which is the expected behavior.

Let me know if this should be opened as a separate issue instead of an additional comment on this one. I'm working to identify when this works versus not. Right now it looks like Ctrl+C works when I don't edit any files (i.e. Ctrl+C right after launching), but editing a file puts watchexec in a bad state, causing it to not respond to file edits OR Ctrl+C.

**Comment by passcod** [comment]
Yeah, 1.17 also printed an os error 10 sometimes in this situation. I thought I'd caught most of it this time round, but it's a pernicious race condition. OTOH, it's doesn't really affect anything, so I might just hide it from view by default.

---

The first log you have there ("When I edit a file") is basically "business as usual." It's catching a file change, the file ceases existing before it can stat it, it emits an event with `file_type: None`, repeat for almost every other file change. It's what happens _after_, or the absence of things happening of nothing is, that would help.

---

The second log you have (hitting ctrl-C) is more worrying, as there should really be something else here. At the very least, it should have a `watchexec::action` sequence there. That it doesn't might mean that crashed earlier and didn't terminate the process, or that it's stuck waiting for something that never happens, or that a channel is dead without erroring...

I'd love to have full logs for these, if you want to email them in or gist/pastebin them or something.

**Comment by kurtbuilds** [comment]
To be clear, those are the full logs after I encountered this. That is to say, in the first example, ("when I edit a file"), it's not triggering a re-run, even though there are events without `file_type: None`. Note that the first two events are for a filename `main.rs`, and they have a value for `file_type`. The latter events are for a editor backup file `main.rs~`, and watchexec gets `file_type: None` when stat-ing those. `watchexec` should be re-triggering when it sees those first events, and it's not.

I'm quite confused because I'm getting extremely inconsistent behavior. When I tried reproducing maybe 20 minutes ago, I got everything working as expected (even that error message wasn't appearing). I then tried again shortly thereafter, and I'm not getting any events shown in trace, either from edit events or from Ctrl+C.

I'm struggling to find anything I'm doing differently between these invocations, nor can I think of anything that's particuarly unique about my environment. (As mentioned, OS X 11.6.2, and mostly modifying files with a recent version of IntelliJ, but I've encountered the same non-responsiveness editing files with command-line vim as well.)

**Comment by passcod** [comment]
Right, I would like the full logs *before* that point, to see if there are any hints prior to then. Additionally, if you could provide a log with `RUST_LOG=trace` set instead of `-vvv`, that may show if e.g. Tokio internals have further clues.

**Comment by kurtbuilds** [comment]
Alright, I managed to reproduce it with traces on. These are just `-vvv`. If you also need RUST_LOG=trace, I can re-record. I truncated to final 1024 lines because each full log is ~24MB and gist supports max 1 MB. See gist here: https://gist.github.com/kurtbuilds/b8b1bc69903f9fe49e627f8948ad87cf

The first log is it working correctly. The second is where `watchexec` was not printing any logs/was non-responsive when hitting Ctrl+C.

That said, after playing with this a bit, it seemed to me like watchexec stops being responsive when it hits the "WARN" lines on the `main.rs~` file. I confirmed there aren't any `warn` lines on the `quits_successfully.log`, and there are `warn` lines associated with `main.rs~` on the failure log.

Let me know if there's more info I can provide to triangulate the issue, and thank you for all your help looking into it!

**Comment by passcod** [comment]
Alright, here's the interesting bit:

```
2022-01-26T23:21:28.128993Z TRACE watchexec::action: out of throttle, starting action process
2022-01-26T23:21:28.129011Z DEBUG watchexec::action: action constructed action=Action { events: [Event { tags: [Source(Filesystem), FileEventKind(Modify(Metadata(Any))), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs", file_type: Some(File) }], metadata: {} }, Event { tags: [Source(Filesystem), FileEventKind(Modify(Data(Content))), Path { path: "/Users/kurt/work/rust-service/server/src/main.rs", file_type: Some(File) }], metadata: {} }], outcome: OnceCell(Uninit) }
2022-01-26T23:21:28.129072Z DEBUG watchexec::action: running action handler
2022-01-26T23:21:28.129174Z DEBUG watchexec::action: handler finished outcome=IfRunning(Both(Wait, Start), Start)
2022-01-26T23:21:28.129193Z TRACE watchexec::command::supervisor: supervisor state ongoing=true
2022-01-26T23:21:28.129203Z DEBUG watchexec::action: outcome resolved outcome=Both(Wait, Start)
2022-01-26T23:21:28.129219Z TRACE watchexec::action: applying outcome outcome=Both(Wait, Start)
2022-01-26T23:21:28.129228Z TRACE watchexec::action: applying outcome outcome=Wait
2022-01-26T23:21:28.129235Z DEBUG watchexec::command::supervisor: waiting on supervisor completion
```

This is followed by the log you gave above.

Basically, your command is still running. You're not passing `-r`, so we don't try to restart it. The subsequent events should still get processed, though... and there's the "bug", it pauses the event processing while waiting for the process to end. That's... possibly not ideal. I'll think about it.

Now, if you're expecting the process to have died in the meantime, that's an issue, but given it's a server I'm guessing that's not the case?

**Comment by passcod** [comment]
Alright, the underlying issue (event processing being paused) is fixed in 1.18.5.

**Comment by kurtbuilds** [comment]
Thank you!
